### PR TITLE
decouple force_on_policy_ratio from step0 reforward.

### PR DESCRIPTION
**Pull request by entrpn**
Measured on qwen3-4b (OpenMathInstruct-2, agentic rollout):
  - clipfrac       0.0006 -> 0.0000
  - log_ratio_abs  0.0157 -> 0.0000
  - step time      984s   -> 742s (~24%)

Gradient magnitude is essentially unchanged; clipping was only firing on
~0.06% of tokens, all of it kernel noise rather than policy drift.

Also in this change:
- log_sampler_trainer_agreement: opt-in flag that spends one trainer
  forward pass to keep the sampler/trainer agreement metrics alive.
  Without it, enabling force_on_policy_ratio silently drops them.
- force_on_policy_ratio + off_policy_steps > 0 now warns instead of
  raising. The MLPerf reference pairs the two deliberately.
- Compare completion_tokens[-1] against EOS when counting clipped
  completions; the mask was being compared against a vocab id, so
  clip_ratio over-counted.
- Buffer metrics once per group instead of once per env/reward-time
  sub-key. Previously duplicated every metric, or dropped all of them
  when no timing keys were present.

Matches the MLPerf Qwen3.5-397B GRPO reference recipe.

**Reference**

Same implementation as nemo_rl for force_on_policy_ratio, found here: https://github.com/NVIDIA-NeMo/RL/blob/main/nemo_rl/algorithms/loss/loss_functions.py#L520C9-L525C36

**Checklist**

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and all unit tests pass.
- [x] I have added all appropriate doc-strings/documentation.
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).
- [x] I have followed [Contribution Guidelines](https://github.com/google/tunix/blob/main/docs/contributing.md).

### CODE REVIEW

File: tests/rl/agentic/agentic_grpo_learner_test.py [hunk 1]
```diff
@@ -2375,6 +2375,185 @@ def _patch_process_results(
...
+      train_example = results[0]
+
+      # 1. Asserts 0 extra trainer forward passes!
+      mock_get_actor_logps.assert_not_called()
+
+      # 2. Asserts old_per_token_logps is None (forces ratio to 1.0 via stop_gradient in loss)
+      self.assertIsNone(train_example.old_per_token_logps)
+
+
+  def test_compute_packed_logps_force_on_policy_skips_recompute(self):
```
Comment (tianshub) [inline tests/rl/agentic/agentic_grpo_learner_test.py]: I guess we can remove this test since logps recompute has nothing to do with seq packing, so we save some testing overhead.

File: tunix/rl/agentic/agentic_grpo_learner.py [hunk 2]
```diff
@@ -124,6 +124,7 @@ class GRPOConfig(agentic_rl_learner.AgenticRLConfig):
   # significantly with the rollout sampler. Without this correction, importance
   # ratios computed directly against the sampler's logp can spike on outlier
   # tokens, producing large-variance gradient updates.
+  force_on_policy_ratio: bool = False
```
Comment (tianshub) [inline tunix/rl/agentic/agentic_grpo_learner.py]: this should be moved before L115 with a comment. e.g. An optimized version of recomputing actor logps when num_iteration equals to 1, instead of recomputing, we will directly use detached logps from actor model.

File: tunix/rl/agentic/agentic_grpo_learner.py [hunk 3]
```diff
@@ -646,7 +661,11 @@ def _process_results(
 
     rollout_per_token_logps = None
     trainer_per_token_logps = None
-    if self.algo_config.use_rollout_logps and padded_old_logprobs:
+    if self.algo_config.force_on_policy_ratio:
+      # Bypass both rollout logprobs and Step-0 Re-Forward passes.
+      # Settting this to None instructs grpo_loss_fn to use stop_gradient
```
Comment (tianshub) [inline tunix/rl/agentic/agentic_grpo_learner.py]: setting, also `setting this to None instructs grpo_loss_fn to use stop_gradient` isn't accurate, the loss fn also depends on the knob value, we can just remove this comment

File: tunix/rl/agentic/agentic_grpo_learner.py [hunk 4]
```diff
@@ -680,6 +699,7 @@ def _process_results(
     elif is_packed:
       old_per_token_logps = None
     else:
+      # Step-0 Re-Forward path (use_rollout_logps=False, force_on_policy_ratio=False)
```
Comment (tianshub) [inline tunix/rl/agentic/agentic_grpo_learner.py]: this is not step-0 only, so we should remove the term

**Review by tianshub — CHANGES_REQUESTED** [review]

File: tunix/rl/algo_core.py [hunk 5]
```diff
@@ -430,7 +430,9 @@ def grpo_loss_fn(
   # TODO(tsbao): We should handle token level advantages.
   advantages = jnp.astype(train_example.advantages, jnp.float32)
 
-  if train_example.old_per_token_logps is None:
+  force_on_policy = getattr(algo_config, "force_on_policy_ratio", False)
+
+  if force_on_policy or train_example.old_per_token_logps is None:
```
Comment (tianshub) [inline tunix/rl/algo_core.py]: actually do you need to check on force_on_policy? since old_per_token_logps won't be populated with force_on_policy

File: tunix/rl/agentic/agentic_grpo_learner.py (hunk 2 above)
Comment (lc5211) [inline tunix/rl/agentic/agentic_grpo_learner.py]: Can we add docstring for force_on_policy_ratio?

File: tunix/rl/algo_core.py (hunk 5 above)
Comment (lc5211) [reply tunix/rl/algo_core.py]: +1, old_per_token_logps is already None with force_on_policy.

File: tests/rl/agentic/agentic_grpo_learner_test.py (hunk 1 above)
Comment (entrpn) [reply tests/rl/agentic/agentic_grpo_learner_test.py]: done.

File: tunix/rl/agentic/agentic_grpo_learner.py (hunk 3 above)
Comment (entrpn) [reply tunix/rl/agentic/agentic_grpo_learner.py]: done.

File: tunix/rl/agentic/agentic_grpo_learner.py (hunk 2 above)
Comment (entrpn) [reply tunix/rl/agentic/agentic_grpo_learner.py]: done.

File: tunix/rl/agentic/agentic_grpo_learner.py (hunk 4 above)
Comment (entrpn) [reply tunix/rl/agentic/agentic_grpo_learner.py]: done.

File: tunix/rl/agentic/agentic_grpo_learner.py (hunk 2 above)
Comment (entrpn) [reply tunix/rl/agentic/agentic_grpo_learner.py]: done.

File: tunix/rl/algo_core.py (hunk 5 above)
Comment (entrpn) [reply tunix/rl/algo_core.py]: good call, I removed it.

**Comment by entrpn** [comment]
@lc5211 @tianshub resolved your comments, PTAL.

File: tunix/rl/agentic/agentic_grpo_learner.py [hunk 6]
```diff
@@ -83,6 +83,17 @@ class GRPOConfig(agentic_rl_learner.AgenticRLConfig):
     max_concurrency: Maximum number of concurrent rollout engines.
     off_policy_steps: Number of off-policy steps can be accepted before a policy
       update.
+    use_rollout_logps: Use the rollout engine's log-probabilities as
+      old_per_token_logps. False makes the trainer recompute them.
+    force_on_policy_ratio: When num_iterations == 1, use
+      stop_gradients(current_logp) as old_per_token_logps instead of recomputing
+      or using rollout logps. Pins the surrogate ratio to 1.0, so clipping never
+      fires and sampler-vs-trainer numerical noise is removed from the ratio.
+    log_sampler_trainer_agreement: Optinally spend one extra trainer forward
```
Comment (tianshub) [inline tunix/rl/agentic/agentic_grpo_learner.py]: Optinally -> Optionally

File: tunix/rl/agentic/agentic_grpo_learner.py [hunk 7]
```diff
@@ -112,6 +123,12 @@ class GRPOConfig(agentic_rl_learner.AgenticRLConfig):
       False  # Whether to mask out degenerate groups with all-0 advantages.
   )
   use_rollout_logps: bool = True
+  # Pin the surrogate ratio to 1.0 (old_logp := stop_gradient(current_logp)).
+  # Valid for single-iteration on-policy training only.
+  force_on_policy_ratio: bool = False
+  # Costs one trainer forward pass; keeps the sampler/trainer agreement metrics
+  # alive when force_on_policy_ratio would otherwise leave nothing to compare.
+  log_sampler_trainer_agreement: bool = False
```
Comment (tianshub) [inline tunix/rl/agentic/agentic_grpo_learner.py:131]: @lc5211 are there any concerns if we disable this? probably no, but just to be aware this is another knob if we want to see the logged metrics

File: tunix/rl/agentic/agentic_grpo_learner.py [hunk 8]
```diff
@@ -573,7 +622,7 @@ def _process_results(
       )
       if (
           len(completion_tokens) >= max_response_length
-          and completion_mask[-1] != eos_value
+          and completion_tokens[-1] != eos_value
```
Comment (tianshub) [inline tunix/rl/agentic/agentic_grpo_learner.py:625]: ah wow, is this a bug before?

File: tunix/rl/agentic/agentic_grpo_learner.py [hunk 9]
```diff
@@ -845,11 +911,11 @@ def _process_results(
             f"{prefix}/{sub_key}/max": (np.max(flat_vals), np.max),
             f"{prefix}/{sub_key}/min": (np.min(flat_vals), np.min),
         })
-        self.rl_engine.buffer_metrics_async(
-            metrics_to_log,  # pyrefly: ignore[bad-argument-type]
-            mode=mode,
-            step=expected_step,  # pyrefly: ignore[bad-argument-type]
-        )
+      self.rl_engine.buffer_metrics_async(
```
Comment (tianshub) [inline tunix/rl/agentic/agentic_grpo_learner.py:914]: why is this changed?

File: tunix/rl/agentic/agentic_grpo_learner.py (hunk 8 above)
Comment (entrpn) [reply tunix/rl/agentic/agentic_grpo_learner.py:625]: yes this was there already. I added the fix.

File: tunix/rl/agentic/agentic_grpo_learner.py (hunk 9 above)
Comment (entrpn) [reply tunix/rl/agentic/agentic_grpo_learner.py:914]: if timedicts was empty, we would lose the upstream metrics if we keep it inside the if statement. Also it just keeps running the update on flat_vals as the for loop updates it which is unnecessary. I can revert if, if you prefer, it doesn't affect the rest of the PR.

File: tunix/rl/agentic/agentic_grpo_learner.py (hunk 6 above)
Comment (entrpn) [reply tunix/rl/agentic/agentic_grpo_learner.py]: fixed.

File: tunix/rl/agentic/agentic_grpo_learner.py (hunk 7 above)
Comment (lc5211) [reply tunix/rl/agentic/agentic_grpo_learner.py:131]: It's used in combination with force_on_policy_ratio. 
```not self.algo_config.force_on_policy_ratio
            or self.algo_config.log_sampler_trainer_agreement``` in this implementation so we are good. maybe let's add comment in line 486 just for the context purpose on when we enforce training logps recomputation.

File: tunix/rl/agentic/agentic_grpo_learner.py (hunk 9 above)
Comment (tianshub) [reply tunix/rl/agentic/agentic_grpo_learner.py:914]: gotcha

**Review by tianshub — APPROVED** [review]
